packages/office-ui-fabric-react/src/components/ContextualMenu/ContextualMenu.Props.ts

```diff
@@ -261,7 +261,7 @@ export interface IContextualMenuItem {
    * Method to custom render this menu item
    * @defaultvalue undefined
    */
-  onRender?: (item: any) => React.ReactNode;
+  onRender?: (item: any, index?: number, hasCheckmarks?: boolean, hasIcons?: boolean) => React.ReactNode;
```

Comment:
I think this is really the wrong direction and we need to consider deprecating this method and replacing it with a better contract that implements `IRenderFunction`. 

First, this list is not complete at all. What about the checked state? What about the onClick that opens a submenu or dismisses the contextual menu? I worked with a partner on implementing a custom ContextualMenuItem and it's severely limited.

The correct design is:

1. Abstract out a ButtonBase variant for ContextualMenuItem.
2. Add onRenderItem: (props: IButtonProps, defaultRender: (props) => JSX.Element)): JSX.Element
3. Add necessary updates to IButtonProps to support this.
4. Deprecate old onRender method.

 We need this to support a variety of things. For example today you can't render a custom item that has a submenu.